pkg/storage/mvcc.go

```diff
@@ -3906,6 +3935,215 @@ func MVCCGarbageCollect(
+	}
+
+	sort.Slice(rangeKeys, func(i, j int) bool {
+		iKey := MVCCRangeKey{StartKey: rangeKeys[i].StartKey, Timestamp: rangeKeys[i].Timestamp}
+		jKey := MVCCRangeKey{StartKey: rangeKeys[j].StartKey, Timestamp: rangeKeys[j].Timestamp}
+		return iKey.Compare(jKey) < 0
+	})
+	// Validate that keys are non-overlapping
+	var lastEndKey roachpb.Key
+	for i, rangeKey := range rangeKeys {
+		if rangeKey.StartKey.Compare(lastEndKey) < 0 {
+			return errors.Errorf("range keys in gc request should be non-overlapping: %s vs %s",
+				rangeKeys[i-1].String(), rangeKey.String())
+		}
+		lastEndKey = rangeKey.EndKey
+	}
+
+	var lastFragmentStart roachpb.Key
+	var pendingRanges []MVCCRangeKey
+	deletePendingFragments := func() error {
+		// Delete what was previously found.
+		if len(pendingRanges) > 0 {
+			for _, r := range pendingRanges {
+				if err := rw.ExperimentalClearMVCCRangeKey(r); err != nil {
+					return err
+				}
+			}
+			pendingRanges = pendingRanges[:0]
+		}
+		return nil
+	}
+
+	// Bound the iterator appropriately for the set of keys we'll be garbage
+	// collecting.
+	iter := rw.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{
+		LowerBound: rangeKeys[0].StartKey,
+		UpperBound: rangeKeys[len(rangeKeys)-1].EndKey,
+		KeyTypes:   IterKeyTypePointsAndRanges,
+	})
+	defer iter.Close()
+
+	var currentKey roachpb.Key
+	for _, rangeKey := range rangeKeys {
+		iter.SeekGE(MVCCKey{Key: rangeKey.StartKey})
+
+		// Iterate within fragment till we find the end of it.
+		ok, err := iter.Valid()
+		if err != nil {
+			return err
+		}
+		if !ok {
+			if err := deletePendingFragments(); err != nil {
+				return err
+			}
+			break
+		}
+		hasP, hasR := iter.HasPointAndRange()
+
+		for ok { // Loop over gc'd range
+			if hasR {
+				span := iter.RangeBounds()
+				fragmentStart := span.Key
+				// Check if we landed on the new fragment set and need to find matching.
+				if !fragmentStart.Equal(lastFragmentStart) {
+					lastFragmentStart = append(lastFragmentStart[:0], fragmentStart...)
+					if err := deletePendingFragments(); err != nil {
+						return err
+					}
+					// Check if we moved past the end of GC range first.
+					if fragmentStart.Compare(rangeKey.EndKey) >= 0 {
+						break
+					}
+					// Find candidate fragments.
+					rangeKeyFragments := iter.RangeKeys()
+					idx := sort.Search(len(rangeKeyFragments), func(i int) bool {
+						return rangeKeyFragments[i].RangeKey.Timestamp.LessEq(rangeKey.Timestamp)
+					})
+					// Copy overlapping parts of range fragments below timestamp into pending.
+					if idx < len(rangeKeyFragments) {
+						stackStart := rangeKey.StartKey
+						if fragmentStart.Compare(rangeKey.StartKey) > 0 {
+							stackStart = fragmentStart.Clone()
+						}
+						stackEnd := rangeKey.EndKey
+						if span.EndKey.Compare(rangeKey.EndKey) < 0 {
+							stackEnd = span.EndKey.Clone()
+						}
+						for ; idx < len(rangeKeyFragments); idx++ {
+							pendingRanges = append(pendingRanges, MVCCRangeKey{
```

Comment:
Do we really need to buffer these? Isn't it sufficient to just clear them as we encounter them?

Reply:
That's valid. I was initially merging key fragments when doing delete, but it doesn't give any benefits we can just remove straight away.